src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -166,41 +166,45 @@ public IODataResourceTypeContext TypeContext
         /// <param name="resource">The resource instance.</param>
         /// <param name="serializationInfo">The serialization info of the resource for writing without model.</param>
         /// <param name="actualEntityType">The edm entity type of the resource</param>
+        /// <param name="requiresId">Whether key properties are required to be returned</param>
         /// <returns>Key value pair array</returns>
         internal static KeyValuePair<string, object>[] GetKeyProperties(
             ODataResourceBase resource,
             ODataResourceSerializationInfo serializationInfo,
-            IEdmEntityType actualEntityType)
+            IEdmEntityType actualEntityType,
+            bool requiresId)
         {
+
             KeyValuePair<string, object>[] keyProperties = null;
-            string actualEntityTypeName = null;
+            string actualEntityTypeName = resource.TypeName ?? actualEntityType?.FullName();

```

Comment:
Do you know how important the old check for empty string is? We are missing that now, maybe this should be `string.IsNullOrEmpty(resource.TypeName) ? actualEntityType?.FullName() : resource.TypeName`

Reply:
It is unlikely this will be an empty string, but I like the robustness of your suggestion and will adopt it.